### DISCUSSION TITLE
Subject contains the first part of the email address before the @

### DISCUSSION THREAD

**Discussion by MuenchenBeiNacht** [rspamd/rspamd#5022] [Ideas]
**Hello RSPAMD gurus and experts.**

A few months ago a perfectly functioning anti-spam solution for ["SPAM emails with the subject changing by one character"](https://github.com/rspamd/rspamd/discussions/4604) was suggested to me and the suggestion implemented back then still filters out hundreds of SPAM emails every day.

I'm now annoyed by the following SPAM problem that I can't solve. 

Maybe someone has an idea that will help me. 

The following new annoyance: 

The subject of SPAM emails always begins with the text before @domaine.com. So for an email address tom@domaine.com the subject is "Tom, rich through AI" or "Tom rich through Bitcoin" - sometimes there is a "," after the name, sometimes not. The texts in the subject vary with a few key words such as: AI, Bitcoin, stocks, etc. 

Is there a rule if the recipient's name appears in the subject, as in the example "Tom" here, which could result in a higher SPAM score? That would generally help.

And can one create a rule with a combination of 1.) the name before the @ appears in the subject and 2.) the subject is followed by one or more terms from a list of keywords such as "AI, Bitcoin, stocks" etc. == >> then bad SPAM = higher score factor?

I am grateful for every suggestion.

Kind regards - Ewald - MuenchenBeiNacht!

**Comment by tobiaskolbe** [comment]
I am receiving such mails as well. If you have only a few recipients on your server, you could catch them with a regex map:
```
/(user1|user2|user3)(,)?( )(bitcoin|invest now)/i
```
If you have a dynamic amount of users on your server, a lua rule could help (/etc/rspamd/rspamd.local.lua):
```
-- Test
local demo_map = rspamd_config:add_map{
	type = "regexp",
	url = "file:///etc/rspamd/local.d/maps/demo.map",
	description = "This is a demo map"
}

local function own_filter(task)
	local subject = task:get_subject();
	if subject and #subject > 0 then
		local rcpt = task:get_recipients(1); -- Check for smtp recipients
		if rcpt and #rcpt == 1 then -- We assume that there's just one recipient (without forwarding) because the subject always includes the user
			rcpt = rcpt[1].user:lower() -- Spammers use "Tom" and "tom"
			if rcpt then
				local _, pos = subject:lower():find("%f[%a]"..rcpt.."%f[%A]")	-- only find "tom" and not "tommy"
												-- see frontier pattern
												-- https://stackoverflow.com/a/31491467
				if pos then
					if #subject > pos then -- user is in subject and there's still text after the user
						subject = subject:sub(pos+1);
						if demo_map:get_key(subject) then
							-- original subject started as sth like "Re: Tom, buy Bitcoins now!"
							-- the cropped one now is ", buy Bitcoins now!"
							-- so regexps like /^buy Bitcoins now!/ won't work
							-- adjust either the code or the ma
							return true;
						end
					end
				end
			end
		end
	end
	return false;
end

rspamd_config:register_symbol{
	name = "FOUND_YA",
	type = "normal",
	score = 5.0,
	description = "Susp. mail has been found!",
	callback = own_filter
}
```
The demo regex map would look like this:
```
/bitcoin/i
/invest now/i
```
This was just quick and dirty, not properly checked. Hopefully it gives you an idea.

**Reply by tobiaskolbe** [reply]
Ok, I was thinking about the find function. It would find `tom@example.com` but not `tom-jones@example.com`. One idea is to escape the hyphen as in [this function](https://github.com/lua-nucleo/lua-nucleo/blob/v0.1.0/lua-nucleo/string.lua#L245-L267).
So we could adjust
```
[...]
rcpt = rcpt[1].user:lower() -- Spammers use "Tom" and "tom"
[...]
```
to
```
[...]
rcpt = rcpt[1].user:lower() -- Spammers use "Tom" and "tom"
if rcpt and type(rcpt) == "string" then rcpt = escape_lua_pattern(rcpt) end
[...]
```
with [escape_lua_pattern](https://github.com/lua-nucleo/lua-nucleo/blob/v0.1.0/lua-nucleo/string.lua#L245-L267) defined somewhere in the file.

Alternatively, instead of using more external code we could use the [regexp module](https://rspamd.com/doc/lua/rspamd_regexp.html#f8965d) of rspamd.
